File: packages/shared-protocol/src/ts/dspark.ts
/**
 * DSpark 推测解码契约 — TS 镜像。
 *
 * 注意：本模块使用 **snake_case**（与 FastAPI 响应 / dspark.json 一致），
 * 不走 shared-protocol 默认的 camelCase 别名约定。原因：
 *   1. DSpark V0 API + 持久化文件 + 前端已全部用 snake_case
 *   2. 加 alias 转换会破坏现有测试与持久化兼容性
 *   3. 类型对齐成本 > 风格统一成本
 *
 * 如果 Phase 14 决定全局改名，DSpark 可以单独走 alias 路径，不影响其它模块。
 *
 * 文档：[docs/design/phase-13-dspark.md](../../../../docs/design/phase-13-dspark.md)
 */

// === 校验上下界常量（前后端共用唯一真源）==================================

export const DSPARK_CONTEXT_SIZE_MIN = 512;
export const DSPARK_CONTEXT_SIZE_MAX = 262144;
export const DSPARK_CONTEXT_SIZE_DEFAULT = 4096;

export const DSPARK_GPU_LAYERS_MIN = -1;
export const DSPARK_GPU_LAYERS_MAX = 999;
export const DSPARK_GPU_LAYERS_DEFAULT = 0;

export const DSPARK_SHORT_OUTPUT_MIN = 1;
export const DSPARK_SHORT_OUTPUT_DEFAULT = 20;

export const DSPARK_N_DRAFT_MIN = 1;
export const DSPARK_N_DRAFT_MAX = 16;

export const DSPARK_DRAFT_P_MIN_DEFAULT_AGGRESSIVE = 0.75;
export const DSPARK_DRAFT_P_MIN_DEFAULT_STANDARD = 0.85;
export const DSPARK_DRAFT_P_MIN_DEFAULT_CONSERVATIVE = 0.9;
export const DSPARK_DRAFT_P_MIN_DEFAULT_OFF = 1.0;

// === 字面量类型 ============================================================

export type SpeculativeMode = 'aggressive' | 'standard' | 'conservative' | 'off';

export type DSparkDecisionReason =
  | 'applied'
  | 'applied-default'
  | 'off-global'
  | 'off-local-only'
  | 'off-short'
  | 'off-no-draft'
  | 'off-no-runtime';

// === 模型 ==================================================================

export interface SpeculativePolicy {
  task_category: string;
  mode: SpeculativeMode;
  n_draft: number;
  draft_p_min: number;
  enabled?: boolean; // 仅 GET /dspark/policies 响应携带，POST body 不需要
}

export interface DSparkConfig {
  draft_model_path: string | null;
  short_output_threshold: number;
  enable_global: boolean;
  context_size: number;
  gpu_layers: number;
}

/** GET /dspark/config 响应：DSparkConfig + 服务端统计 + 路径信息 */
export interface DSparkRuntimeConfig extends DSparkConfig {
  yaml_path: string | null;
  profile_count: number;
  stats: {
    total_decisions: number;
    dspark_enabled_pct: number;
    per_category: Record<string, number>;
    per_reason: Record<string, number>;
  };
}

/** GET /dspark/recent 单条记录（engine deque，UI 必须显示"本次会话"）*/
export interface DSparkDecisionRecord {
  ts: number;
  task_category: string;
  speculative_enabled: boolean;
  n_draft: number;
  draft_p_min: number;
  backend: string;
  reason: DSparkDecisionReason;
  max_tokens: number;
}

/** POST /dspark/config 请求体（所有字段可选，只更新非 null 字段）*/
export interface DSparkConfigUpdateBody {
  draft_model_path?: string | null;
  short_output_threshold?: number;
  enable_global?: boolean;
  context_size?: number;
  gpu_layers?: number;
}

/** POST /dspark/draft-model-path 请求体 */
export interface DSparkDraftModelPathBody {
  path: string | null;
}

// === 工具 ==================================================================

/** Mode → 默认 K/阈值映射（4 档预设） */
export const MODE_PARAMS: Record<SpeculativeMode, { n_draft: number; draft_p_min: number }> = {
  aggressive: {
    n_draft: 8,
    draft_p_min: DSPARK_DRAFT_P_MIN_DEFAULT_AGGRESSIVE,
  },
  standard: {
    n_draft: 4,
    draft_p_min: DSPARK_DRAFT_P_MIN_DEFAULT_STANDARD,
  },
  conservative: {
    n_draft: 2,
    draft_p_min: DSPARK_DRAFT_P_MIN_DEFAULT_CONSERVATIVE,
  },
  off: {
    n_draft: 1,
    draft_p_min: DSPARK_DRAFT_P_MIN_DEFAULT_OFF,
  },
};

/** 决策原因 → 颜色（与 Python REASON_COLOR 一一对应） */
export const REASON_COLOR: Record<DSparkDecisionReason, string> = {
  applied: '#4ec9b0',
  'applied-default': '#dcdcaa',
  'off-global': '#858585',
  'off-local-only': '#f48771',
  'off-short': '#dcdcaa',
  'off-no-draft': '#858585',
  'off-no-runtime': '#f48771',
};

/** 模式 → 颜色（4 档配色） */
export const MODE_COLOR: Record<SpeculativeMode, string> = {
  aggressive: '#dcdcaa',
  standard: '#569cd6',
  conservative: '#4ec9b0',
  off: '#858585',
};

File: packages/shared-protocol/src/protocol/dspark.py
"""DSpark 推测解码契约（Pydantic 镜像）。

跨语言协议类型：TS ⇄ Rust ⇄ FastAPI ⇄ MCP servers。
本文件定义所有 DSpark 配置的字段 + 校验上下界，前后端共用。

校验上下界是 Phase 13 V1 llama.cpp 实际取值范围（设计文档 §3.1）。
前端 DSparkSettingsPanel 不得私自收紧，否则多客户端行为不一致（问题 4）。

文档：[docs/design/phase-13-dspark.md](../../../docs/design/phase-13-dspark.md)
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


# === 校验边界常量（前后端共用唯一真源） ======================================

# 上下文窗口大小
DSPARK_CONTEXT_SIZE_MIN = 512
DSPARK_CONTEXT_SIZE_MAX = 262144
DSPARK_CONTEXT_SIZE_DEFAULT = 4096

# GPU 层数（llama.cpp n_gpu_layers）
DSPARK_GPU_LAYERS_MIN = -1  # -1 = 自动全部
DSPARK_GPU_LAYERS_MAX = 999
DSPARK_GPU_LAYERS_DEFAULT = 0  # 纯 CPU

# 短输出阈值
DSPARK_SHORT_OUTPUT_MIN = 1
DSPARK_SHORT_OUTPUT_DEFAULT = 20

# 推测解码 K（每轮草稿 token 数）
DSPARK_N_DRAFT_MIN = 1
DSPARK_N_DRAFT_MAX = 16

# 草稿模型接受概率阈值
DSPARK_DRAFT_P_MIN_DEFAULT_AGGRESSIVE = 0.75
DSPARK_DRAFT_P_MIN_DEFAULT_STANDARD = 0.85
DSPARK_DRAFT_P_MIN_DEFAULT_CONSERVATIVE = 0.90
DSPARK_DRAFT_P_MIN_DEFAULT_OFF = 1.0


# === 字面量类型 ==============================================================

# 4 档预设（设计文档 §2.2 置信度阈值对照表）
SpeculativeMode = Literal["aggressive", "standard", "conservative", "off"]

# 决策原因常量（5 关铁律的输出 reason）
# TS 端 DSparkDecisionReason 必须与此处一一对应。
DSparkDecisionReason = Literal[
    "applied",
    "applied-default",
    "off-global",
    "off-local-only",
    "off-short",
    "off-no-draft",
    "off-no-runtime",
]


# === 模型 ====================================================================


class SpeculativePolicy(BaseModel):
    """单任务类别的 DSpark 策略。

    `mode` 决定 4 档预设模板（aggressive / standard / conservative / off）；
    也可直接覆盖 n_draft / draft_p_min（YAML `profiles.<name>.n_draft`）。
    """

    task_category: str = Field(min_length=1)
    mode: SpeculativeMode = "off"
    n_draft: int = Field(
        default=1,
        ge=DSPARK_N_DRAFT_MIN,
        le=DSPARK_N_DRAFT_MAX,
    )
    draft_p_min: float = Field(default=1.0, ge=0.0, le=1.0)

    @property
    def enabled(self) -> bool:
        """off 模式永远 disabled；其它模式要求 n_draft > 1。

        draft_p_min >= 1.0 也视为关闭（接受率为 1.0 = 无确认需求 = 没意义）。
        """
        if self.mode == "off":
            return False
        if self.n_draft <= 1:
            return False
        if self.draft_p_min >= 1.0:
            return False
        return True


class DSparkConfig(BaseModel):
    """DSpark 全局配置。

    V0 持久化到 `dspark.json`；V1 接 llama.cpp 时按本模型加载。
    `profiles` 字段是业务默认表（在业务层 `dspark/config.py` 添加），
    不属于跨语言协议层。
    """

    draft_model_path: str | None = Field(
        default=None,
        description="草稿模型 GGUF 路径；None 或空串 = 全局禁用 DSpark",
    )
    short_output_threshold: int = Field(
        default=DSPARK_SHORT_OUTPUT_DEFAULT,
        ge=DSPARK_SHORT_OUTPUT_MIN,
        description="max_tokens < 此值时跳过 DSpark（避免负优化）",
    )
    enable_global: bool = Field(
        default=True,
        description="总开关；False 时所有任务类别都关闭 DSpark",
    )
    context_size: int = Field(
        default=DSPARK_CONTEXT_SIZE_DEFAULT,
        ge=DSPARK_CONTEXT_SIZE_MIN,
        le=DSPARK_CONTEXT_SIZE_MAX,
        description="上下文窗口大小（tokens）",
    )
    gpu_layers: int = Field(
        default=DSPARK_GPU_LAYERS_DEFAULT,
        ge=DSPARK_GPU_LAYERS_MIN,
        le=DSPARK_GPU_LAYERS_MAX,
        description="GPU 层数：0=纯 CPU，-1=自动全部，N>0=指定层数",
    )


class DSparkDecisionRecord(BaseModel):
    """单次决策记录（写入 engine deque，UI 通过 GET /dspark/recent 读）。

    V0 内存存储，重启归零（UI 文案必须显示"本次会话"而非"累计"）。
    """

    ts: float = Field(description="决策时间戳（unix epoch seconds）")
    task_category: str
    speculative_enabled: bool
    n_draft: int
    draft_p_min: float
    backend: str
    reason: DSparkDecisionReason
    max_tokens: int